### PR TITLE
feat(anim): add a pause method

### LINKED ISSUE #7029: lv anim support pause/resume
### Problem to solve

anim requires pause/resume in some cases to keep continuity

### Success criteria

pause the anim, sleep several seconds, the resume it started with the same anim value.

### Solution outline

add new control API, like lv_anim_pause/resume

### Rabbit holes

No

### Testing

pause the anim, sleep several seconds, the resume it started with the same anim value.

### Teaching

Doc the new API

### Considerations

_No response_

### PR DESCRIPTION
**Pull request by AndreCostaaa**
Closes #7029 

This PR adds the ability to pause and resume animations

### CODE REVIEW

File: src/misc/lv_anim.c [hunk 1]
```diff
@@ -499,6 +500,30 @@ uint32_t lv_anim_resolve_speed(uint32_t speed_or_time, int32_t start, int32_t en
     return LV_CLAMP(min_time * 10, time, max_time * 10);
 }
 
+bool lv_anim_is_paused(lv_anim_t * a)
+{
+    return a->is_paused;
+}
+
+void lv_anim_pause(lv_anim_t * a)
```
Comment (uLipe) [inline src/misc/lv_anim.c:510]: What happens if `a` is NULL? Is guaranteed this will never occur? Additionally since this a function that does not return value, I would like to see at least a LOG_ERR/WRN when things get wrong, it would facillitate and avoid silent failures.

File: src/misc/lv_anim.c [hunk 2]
```diff
@@ -499,6 +500,30 @@ uint32_t lv_anim_resolve_speed(uint32_t speed_or_time, int32_t start, int32_t en
...
+{
+    lv_anim_pause_for(a, LV_ANIM_PAUSE_FOREVER);
+}
+
+void lv_anim_pause_for(lv_anim_t * a, uint32_t ms)
```
Comment (uLipe) [inline src/misc/lv_anim.c:516]: Please check also here.

File: src/misc/lv_anim.c [hunk 3]
```diff
@@ -499,6 +500,30 @@ uint32_t lv_anim_resolve_speed(uint32_t speed_or_time, int32_t start, int32_t en
...
+{
+    a->is_paused = true;
+    a->pause_time = lv_tick_get();
+    a->pause_duration = ms;
+}
+
+void lv_anim_resume(lv_anim_t * a)
```
Comment (uLipe) [inline src/misc/lv_anim.c:522]: And here.

**Review by uLipe — CHANGES_REQUESTED** [review]
Nice work! I just have a small concern pointed, could you check please?

File: src/misc/lv_anim.c (hunk 1 above)
Comment (AndreCostaaa) [reply src/misc/lv_anim.c:510]: If we really want  to make sure that `a` is valid, we would need to go through the animation list to find `a`.

Having said that I agree with checking if it's NULL as it's always a good practice when dealing with pointers

Comment (uLipe) [reply src/misc/lv_anim.c:510]: There are some asserting macros for NULL you can use on these cases, my intention here is just to provide some hint for the user  if something goes wrong, making easier to debug when the test coverage does not reach some potential edge case.

File: src/misc/lv_anim.c [hunk 4]
```diff
@@ -694,6 +696,13 @@ static int32_t lv_anim_path_cubic_bezier(const lv_anim_t * a, int32_t x1, int32_
     return new_value;
 }
 
+static void lv_anim_pause_for_internal(lv_anim_t * a, uint32_t ms){
```
Comment (uLipe) [inline src/misc/lv_anim.c]: Nit: brackets should be opened on the next line.

Comment (AndreCostaaa) [reply src/misc/lv_anim.c]: Yep, forgot to run the formatting script before the commit. You can check the latest commit

File: src/misc/lv_anim.c (hunk 1 above)
Comment (AndreCostaaa) [reply src/misc/lv_anim.c:510]: Yep, I agree with you

**Review by uLipe — DISMISSED** [review]

File: src/misc/lv_anim.h [hunk 5]
```diff
@@ -146,6 +147,9 @@ struct _lv_anim_t {
 
     /* Animation system use these - user shouldn't set */
     uint32_t last_timer_run;
+    bool is_paused;                           /**<Indicates that the animation is paused */
```
Comment (kisvegabor) [inline src/misc/lv_anim.h]: To save a few bytes please move it above `uint8_t reverse_play_in_progress : 1;`
```suggestion
    uint8_t is_paused :1;                           /**<Indicates that the animation is paused */
```

**Review by kisvegabor — COMMENTED** [review]
Looks good to me, just on note.

**Review by uLipe — DISMISSED** [review]

File: src/misc/lv_anim.c [hunk 6]
```diff
@@ -519,8 +546,13 @@ static void anim_timer(lv_timer_t * param)
 
     while(a != NULL) {
         uint32_t elaps = lv_tick_elaps(a->last_timer_run);
-        a->act_time += elaps;
 
+        if(a->is_paused) {
+            a->is_paused = (lv_tick_get() - a->pause_time) < a->pause_duration;
```
Comment (liamHowatt) [inline src/misc/lv_anim.c]: Would this be applicable?

```suggestion
            uint32_t time_spent_paused = lv_tick_get() - a->pause_time;
            if(time_spent_paused >= a->pause_duration) {
                a->is_paused = false;
                uint32_t pause_overrun = time_spent_paused - a->pause_duration;
                a->act_time += pause_overrun;
            }
```

Comment (liamHowatt) [reply src/misc/lv_anim.c]: Also, maybe instead of `lv_tick_get() - a->pause_time`, `lv_tick_elaps(a->pause_time)` could be used to handle tick overflow wrap-around as used by `elaps`.

Comment (AndreCostaaa) [reply src/misc/lv_anim.c]: Wouldn't updating the `act_time` mean a small "jump" when the animation is resumed ? It would probably go unnoticed as long as the `pause_overrun` is not too big but it got me wondering

Comment (liamHowatt) [reply src/misc/lv_anim.c]: There would be a jump, ~~but there would be an even bigger jump at the next refresh without it~~. Here's a test case that passes with the change and fails without

```c
void test_anim_pause_for(void)
{
    int32_t var;

    /*Start an animation*/
    lv_anim_t a;
    lv_anim_init(&a);
    lv_anim_set_var(&a, &var);
    lv_anim_set_values(&a, 0, 100);
    lv_anim_set_exec_cb(&a, exec_cb);
    lv_anim_set_duration(&a, 100);
    lv_anim_t * animation = lv_anim_start(&a);

    lv_anim_pause_for(animation, 20);

    lv_test_wait(40);
    TEST_ASSERT_EQUAL(19, var);
}
```

Comment (liamHowatt) [reply src/misc/lv_anim.c]: Also this more important issue. The overrun is lost and not counted. In fairness, it would only be noticeable if the refresh rate was very low.

```c
    ...
    lv_anim_t * animation = lv_anim_start(&a);

    lv_anim_pause_for(animation, 20);

    lv_test_wait(40);
    lv_test_wait(40);
    TEST_ASSERT_EQUAL(59, var); // "Expected 59 Was 39"
}
```

Comment (AndreCostaaa) [reply src/misc/lv_anim.c]: Thank you for providing these test cases and i agree with you, will add the test cases and update the code. Nice catch !

**Review by liamHowatt — APPROVED** [review]
Thanks! Nice catch with the `FOREVER`.

**Review by kisvegabor — APPROVED** [review]
Thank you!
